execution/transaction-orderer/src/batch_orderer.rs

```diff
@@ -0,0 +1,381 @@
+    }
+
+    fn commit_prefix_callback<F, R>(&mut self, count: usize, callback: F) -> R
+    where
+        F: FnOnce(Vec<Self::Txn>) -> R,
+    {
+        assert!(count <= self.count_selected());
+
+        let committed_indices: Vec<_> = (0..count)
+            .map(|_| self.selected.pop_first().unwrap().1)
+            .collect();
+
+        let committed_txns: Vec<_> = committed_indices
+            .iter()
+            .map(|&idx| self.txn_info[idx].transaction.clone())
+            .collect();
+
+        // Return the committed transactions early via the callback, to minimize latency.
+        // Note that the callback cannot access the Orderer as we are still holding a mutable
+        // reference to it. Hence, it will not be able to observe the orderer in an inconsistent
+        // state.
+        let res = callback(committed_txns);
+
+        // Update the internal data structures.
+        self.active_txns_count -= count;
+
+        for &committed_idx in committed_indices.iter() {
+            let tx = &self.txn_info[committed_idx].transaction;
+
+            if let Some(window) = self.window.as_option_mut() {
+                window.recently_committed_txns.push_back(committed_idx);
+                for key in tx.write_set() {
+                    window.recent_writes.entry(key.clone()).or_default().count += 1;
+                }
+            }
+
+            let satisfied_write_table_requests = self
+                .write_reservations
+                .remove_reservations(committed_idx, tx.write_set());
+
+            for (idx, key) in satisfied_write_table_requests {
+                self.satisfy_pending_write_table_request(idx, &key);
+            }
+        }
+
+        // The read table requests have to be processed after all the write table requests,
+        // are processed to make sure that `pending_recent_write_dependencies` is updated.
+        for &committed_idx in committed_indices.iter() {
+            let tx = &self.txn_info[committed_idx].transaction;
+
+            let satisfied_read_table_requests = self
+                .read_reservations
+                .remove_reservations(committed_idx, tx.read_set());
+
+            for (idx, _) in satisfied_read_table_requests {
+                self.satisfy_pending_read_table_request(idx);
+            }
+        }
+
+        res
+    }
+}
+
+impl<T> BatchOrdererWithWindow for SequentialDynamicAriaOrderer<T, ConstSome<WindowManager<T::Key>>>
+where
+    T: PTransaction + Clone,
+    T::Key: Ord + Hash + Eq + Clone,
+{
+    fn forget_prefix(&mut self, count: usize) {
+        assert!(count <= self.get_window_size());
+        let forgotten_indices = self
+            .window
+            .recently_committed_txns
+            .drain(0..count)
+            .collect::<Vec<_>>();
+
+        for forgotten_idx in forgotten_indices {
+            let tx = &self.txn_info[forgotten_idx].transaction;
+            let write_set: Vec<_> = tx.write_set().cloned().collect();
```

Comment:
seems like you don't need to collect, and can just iterate below?
or do .map / for_each